pvlib/iotools/pvgis.py

```diff
@@ -23,6 +23,353 @@
+    data = data.drop('time', axis=1)
+    data = data.astype(dtype={'Int': 'int'})  # The 'Int' column to be integer
+    if map_variables:
+        data.rename(columns=VARIABLE_MAP, inplace=True)
+    return data, inputs, metadata
+
+
+def _parse_pvgis_hourly_csv(src, map_variables):
+    # The first 4 rows are latitude, longitude, elevation, radiation database
+    inputs = {}
+    # 'Latitude (decimal degrees): 45.000\r\n'
+    inputs['latitude'] = float(src.readline().split(':')[1])
+    # 'Longitude (decimal degrees): 8.000\r\n'
+    inputs['longitude'] = float(src.readline().split(':')[1])
+    # Elevation (m): 1389.0\r\n
+    inputs['elevation'] = float(src.readline().split(':')[1])
+    # 'Radiation database: \tPVGIS-SARAH\r\n'
+    inputs['radiation_database'] = str(src.readline().split(':')[1]
+                                       .replace('\t', '').replace('\n', ''))
+    # Parse through the remaining metadata section (the number of lines for
+    # this section depends on the requested parameters)
+    while True:
+        line = src.readline()
+        if line.startswith('time,'):  # The data header starts with 'time,'
+            # The last line of the metadata section contains the column names
+            names = line.replace('\n', '').replace('\r', '').split(',')
+            break
+        # Only retrieve metadata from non-empty lines
+        elif (line != '\n') & (line != '\r\n'):
+            inputs[line.split(':')[0]] = str(line.split(':')[1]
+                                             .replace('\n', '')
+                                             .replace('\r', '').strip())
+    # Save the entries from the data section to a list, until an empty line is
+    # reached an empty line. The length of the section depends on the request
+    data_lines = []
+    while True:
+        line = src.readline()
+        if (line == '\n') | (line == '\r\n'):
+            break
+        else:
+            data_lines.append(line.replace('\n', '').replace('\r', '')
+                              .split(','))
+    data = pd.DataFrame(data_lines, columns=names)
+    data.index = pd.to_datetime(data['time'], format='%Y%m%d:%H%M', utc=True)
+    data = data.drop('time', axis=1)
+    if map_variables:
+        data.rename(columns=VARIABLE_MAP, inplace=True)
+    # All columns should have the dtype=float, except 'Int' which should be
+    # integer. It is necessary to convert to float, before converting to int
+    data = data.astype(float).astype(dtype={'Int': 'int'})
+    # Generate metadata dictionary containing description of parameters
+    metadata = {}
+    for line in src.readlines():
+        if ':' in line:
+            metadata[line.split(':')[0]] = line.split(':')[1].strip()
+    return data, inputs, metadata
+
+
+def read_pvgis_hourly(filename, map_variables=True, pvgis_format=None):
+    """Read a PVGIS hourly file.
+
+    Parameters
+    ----------
+    filename : str, pathlib.Path, or file-like buffer
```

Comment:
Having the `read_` function be able to take a buffer goes against the `get/read/parse` pattern we sort of follow, but I might like this way better.  I'm curious what other people think about it.

Reply:
I copied this functionality from the existing `read_pvgis_tmy`

Reply:
Yeah, like how `pd.read_csv` can take a filepath or a buffer; you don't have to switch to `pd.parse_csv` for when you want to pass a file-like object instead of a filename.  Also less clutter in the API listings with the "two-in-one" approach.  Thoughts from @wholmgren or anyone else?

Reply:
I agree that it's a nicer interface for users. I am less excited about the gymnastics we have to do to support it. The `try/except AttributeError` pattern below feels lacking for some reason. Maybe I'd like it more if we used `hasattr` - it's more explicit, but it does go against the python pattern of preferring try/except to checks. That would also let you avoid repeated calls to the pvlib helper function. In any case, how do other libraries handle this?

Reply:
Without thinking too much about it, one clean approach might be a decorator that does something like (a much simpler version of) pandas's `_get_filepath_or_buffer`: https://github.com/pandas-dev/pandas/blob/master/pandas/io/common.py#L226

Reply:
I'm not crazy about decorators b/c IMHO the abstraction can make the code hard to read and maintain. However, if it's a pattern that we can reuse a lot, then I do think it's worth the effort.

I prefer EAFP try-except vs type-checking, which I think is more Pythonic b/c an object might be able to do the action even if it isn't a subclass of any of the types in the check, but as Emerson says, "A foolish consistency..." so if there's a simpler, cleaner way than try-except, then I'll be in favor.
>Simple is better than complex.
>Complex is better than complicated.

The reason I broke with the get/read/parse pattern for PVGIS was because it had the 4 types, and the number of parsing functions seemed more complicated then just combining read & parse together.

A pattern I usually like for this is recursion. Not sure why I didn't use it here, something like:

```python
def read_fbuf_or_filename(f):
    try:
        f.seek(0)  # or some other method only fbuf has
    except AttributeError:
        # str and Path don't have read
        with pathlib.Path(f).open() as g:
            # now that we have a buffer use recursion to parse
            return read_fbuf_or_filename(g)
    return parse_fbuf(f)
```

I think the reason I didn't like this here is because I wanted to use pandas `to_blah` and the EPW methods which work on file buffers?